jax/_src/scipy/special.py

```diff
@@ -1010,3 +1011,78 @@ def lpmn_values(m: int, n: int, z: jnp.ndarray, is_normalized: bool) -> jnp.ndar
   l_max = n
 
   return _gen_associated_legendre(l_max, z, is_normalized)
+
+
+
+@partial(jit, static_argnums=(4,))
+def _sph_harm(m: jnp.ndarray,
+              n: jnp.ndarray,
+              theta: jnp.ndarray,
+              phi: jnp.ndarray,
+              n_max: int) -> jnp.ndarray:
+  """Computes the spherical harmonics."""
+
+  len_inputs = m.shape[0]
+  harmonics = jnp.zeros((len_inputs), dtype=jnp.complex64)
+  cos_colatitude = jnp.cos(phi)
+
+  legendre = _gen_associated_legendre(n_max, cos_colatitude, True)
+  legendre_val = legendre[abs(m), n, jnp.arange(len(n))]
+
+  angle = abs(m) * theta
+  vandermonde = lax.complex(jnp.cos(angle), jnp.sin(angle))
+  harmonics = lax.complex(legendre_val * jnp.real(vandermonde),
+                          legendre_val * jnp.imag(vandermonde))
+
+  # Negative order.
+  harmonics = jnp.where(m < 0,
+                        (-1.0)**abs(m) * jnp.conjugate(harmonics),
+                        harmonics)
+
+  return harmonics
+
+
+def sph_harm(m: jnp.ndarray,
+             n: jnp.ndarray,
+             theta: jnp.ndarray,
+             phi: jnp.ndarray,
+             n_max: Optional[int] = None) -> jnp.ndarray:
+  r"""Computes the spherical harmonics.
+
+  The JAX version has one extra argument `n_max`, the maximum value in `n`.
+
+  The spherical harmonic of degree `n` and order `m` can be written as
+  :math:`Y_n^m(\theta, \phi) = N_n^m * P_n^m(\cos \phi) * \exp(i m \theta)`,
+  where :math:`N_n^m = \sqrt{\frac{\left(2n+1\right) \left(n-m\right)!}
+  {4 \pi \left(n+m\right)!}}` is the normalization factor and :math:`\phi` and
+  :math:\theta` are the colatitude and longitude, repectively. :math:`N_n^m` is
+  chosen in the way that the spherical harmonics form a set of orthonormal basis
+  functions of :math:`L^2(S^2)`.
+
+  Args:
+    m: The order of the harmonic; must have `|m| <= n`. Return values for
+      `|m| > n` ara undefined.
+    n: The degree of the harmonic; must have `n >= 0`. The standard notation for
+      degree in descriptions of spherical harmonics is `l (lower case L)`. We
+      use `n` here to be consistent with `scipy.special.sph_harm`. Return
+      values for `n < 0` are undefined.
+    theta: The azimuthal (longitudinal) coordinate; must be in [0, 2*pi].
+    phi: The polar (colatitudinal) coordinate; must be in [0, pi].
+    n_max: The maximum degree `max(n)`. If the supplied `n_max` is not the true
+      maximum value of `n`, the results are clipped to `n_max`. For example,
+      `sph_harm(m=jnp.array([2]), n=jnp.array([10]), theta, phi, n_max=6)`
+      acutually returns
+      `sph_harm(m=jnp.array([2]), n=jnp.array([6]), theta, phi, n_max=6)`
+  Returns:
+    A 1D array containing the spherical harmonics at (m, n, theta, phi).
+  """
+  if any(arg.shape != n.shape for arg in [m, theta, phi]):
+    raise ValueError('Not all inputs 1D array have the same length!')
```

Comment:
You'd also have to remove this check to support broadcasted inputs.

Reply:
Done.